File: tests/e2e/harness/delta_diff_engine.py
"""
Live Delta-Diff Engine for Real-Time Streaming Speech Recognition.
Computes Longest Common Prefix (LCP), backspaces needed, and suffix to append,
with full 32-bit Unicode Code Point safety.
"""
from dataclasses import dataclass
from typing import List, Tuple


@dataclass(frozen=True)
class DiffResult:
    common_prefix_length: int  # in Unicode code points
    backspaces_needed: int     # number of backspace keystrokes
    text_to_append: str        # newly added suffix string


class DeltaDiffEngine:
    """
    Computes mathematical diff between old text and new speech hypothesis.
    """

    @staticmethod
    def _to_code_points(text: str) -> List[int]:
        """Converts string into list of 32-bit integer Unicode code points."""
        return [ord(c) for c in text]

    @staticmethod
    def _from_code_points(points: List[int]) -> str:
        """Reconstructs Unicode string from list of integer code points."""
        return "".join(chr(p) for p in points)

    def compute_diff(self, old_text: str, new_text: str) -> DiffResult:
        """
        Computes LCP, backspaces needed, and text to append.
        Operates strictly on Unicode code points to preserve multi-byte emoji integrity.
        """
        old_points = self._to_code_points(old_text)
        new_points = self._to_code_points(new_text)

        lcp = 0
        min_len = min(len(old_points), len(new_points))

        while lcp < min_len and old_points[lcp] == new_points[lcp]:
            lcp += 1

        backspaces_needed = len(old_points) - lcp
        append_points = new_points[lcp:]
        text_to_append = self._from_code_points(append_points)

        return DiffResult(
            common_prefix_length=lcp,
            backspaces_needed=backspaces_needed,
            text_to_append=text_to_append
        )

File: app/src/main/java/com/transcriptor/hid/engine/DeltaDiffEngine.kt
package com.transcriptor.hid.engine

/**
 * Result of a differential computation between the currently rendered host text
 * and an updated speech transcription hypothesis.
 *
 * @property commonPrefixLength Number of matching Unicode code points in the longest common prefix (LCP).
 * @property backspacesNeeded Number of backspace keystrokes required to erase the non-matching suffix.
 * @property textToAppend The new suffix string to append.
 */
data class DiffResult(
    val commonPrefixLength: Int,
    val backspacesNeeded: Int,
    val textToAppend: String
)

/**
 * Interface for computing diffs between consecutive text states.
 */
interface DeltaDiffEngine {
    /**
     * Computes the minimal backspace and append operations using Unicode code points.
     */
    fun computeDiff(oldText: String, newText: String): DiffResult

    companion object {
        fun create(): DeltaDiffEngine = DefaultDeltaDiffEngine()
    }
}

/**
 * Default implementation of [DeltaDiffEngine] operating on 32-bit Unicode code points
 * to ensure grapheme and surrogate pair safety.
 */
class DefaultDeltaDiffEngine : DeltaDiffEngine {

    override fun computeDiff(oldText: String, newText: String): DiffResult {
        if (oldText == newText) {
            val codePointCount = if (oldText.isEmpty()) 0 else oldText.codePointCount(0, oldText.length)
            return DiffResult(
                commonPrefixLength = codePointCount,
                backspacesNeeded = 0,
                textToAppend = ""
            )
        }

        if (oldText.isEmpty()) {
            return DiffResult(
                commonPrefixLength = 0,
                backspacesNeeded = 0,
                textToAppend = newText
            )
        }

        if (newText.isEmpty()) {
            return DiffResult(
                commonPrefixLength = 0,
                backspacesNeeded = oldText.codePointCount(0, oldText.length),
                textToAppend = ""
            )
        }

        val oldPoints = oldText.codePoints().toArray()
        val newPoints = newText.codePoints().toArray()

        var k = 0
        val minLen = minOf(oldPoints.size, newPoints.size)
        while (k < minLen && oldPoints[k] == newPoints[k]) {
            k++
        }

        val backspacesNeeded = oldPoints.size - k
        val appendPoints = newPoints.sliceArray(k until newPoints.size)
        val textToAppend = String(appendPoints, 0, appendPoints.size)

        return DiffResult(
            commonPrefixLength = k,
            backspacesNeeded = backspacesNeeded,
            textToAppend = textToAppend
        )
    }
}
